File: UserService/src/main/scala/users/domain/UserRepository.scala
package me.alstepan.users.domain

import cats.data.EitherT
import me.alstepan.users.domain.Errors.Errors

trait UserRepository[F[_]] {
  def create(user: User): EitherT[F, Errors, User]
  def update(user: User): EitherT[F, Errors, User]
  def delete(id: Long): EitherT[F, Errors, Unit]
  def get(id: Long): EitherT[F, Errors, User]
  def list(): F[List[User]]
  def checkAvailability(): EitherT[F, Errors, Unit]
}

File: UserService-Helidon/src/main/java/me/alstepan/users/domain/UserRepository.java
package me.alstepan.users.domain;

import java.util.List;
import java.util.concurrent.CompletionStage;

public interface UserRepository {
    CompletionStage<User> get(long id);

    CompletionStage<Long> create(User user);

    CompletionStage<Long> delete(long id);

    CompletionStage<Long> update(long id, User user);

    CompletionStage<List<User>> list();
}
